odc/geo/_xr_interop.py

```diff
@@ -245,6 +246,96 @@ def assign_crs(
     return xx
 
 
+def mask(xx: XrT, poly: Geometry, all_touched: bool = True) -> XrT:
+    """
+    Apply a polygon geometry as a mask, setting all xr.Dataset
+    or xr.DataArray pixels outside the rasterized polygon to ``NaN``.
+
+    :param xx:
+       :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`.
+
+    :param poly:
+       Geometry shape used to mask ``xx``.
+
+    :param all_touched:
+        If ``True``, all pixels touched by ``poly`` will remain unmasked.
+        If ``False``, only pixels whose center is within the polygon or
+        that are selected by Bresenham's line algorithm will remain unmasked.
+
+    :return:
+        A :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`
+        masked by ``poly``.
+    """
+
+    # Reproject `poly` to match `xx`
+    poly = poly.to_crs(crs=xx.odc.crs)
+
+    # Rasterise `poly` into geobox of `xx`
+    rasterized = rasterize(
+        poly=poly,
+        how=xx.odc.geobox,
+        all_touched=all_touched,
+    )
+
+    # Mask data outside rasterized `poly`
+    xx_masked = xx.where(rasterized.data)
+
+    return xx_masked
+
+
+def crop(
+    xx: XrT, poly: Geometry, apply_mask: bool = True, all_touched: bool = True
+) -> XrT:
+    """
+    Crops and optionally mask an xr.Dataset or xr.DataArray (``xx``) to
+    the spatial extent of a geometry (``poly``).
+
+    :param xx:
+       :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`.
+
+    :param poly:
+       Geometry shape used to crop ``xx``.
+
+    :param apply_mask:
+       Whether to mask out pixels outside of the rasterized extent of
+       ``poly`` by setting them to ``NaN``.
+
+    :param all_touched:
+        If ``True``, all pixels touched by ``poly`` will remain unmasked.
+        If ``False``, only pixels whose center is within the polygon or
+        that are selected by Bresenham's line algorithm will remain unmasked.
+
+    :return:
+        A :py:class:`~xarray.Dataset` or :py:class:`~xarray.DataArray`
+        cropped and optionally masked to the spatial extent of ``poly``.
+    """
+    # Reproject `poly` to match `xx`
+    poly = poly.to_crs(crs=xx.odc.crs)
+
+    # Verify that `poly` overlaps with `xx` extent
+    if not poly.intersects(xx.odc.geobox.extent):
+        raise ValueError(
+            "The supplied `poly` must overlap spatially with the extent of `xx`."
+        )
```

Comment:
Same story as before: it's dangerous to perform coordinate transform, if it can be avoided by pushing into more basic operations like `.enclosing`, then it should be. You'd then check if overlap_roi returns a null set and raise error then.

Reply:
OK, have revised to remove the coordinate transform, and simply run `enclosing` > `overlap_roi` then verify the ROI isn't empty using `roi_is_empty`:

https://github.com/opendatacube/odc-geo/blob/crop/odc/geo/_xr_interop.py#L317-L328